deepchem/data/dft_data.py

```diff
@@ -0,0 +1,319 @@
+
+    References
+    ----------
+    Kasim, Muhammad F., and Sam M. Vinko. "Learning the exchange-correlation functional from nature with fully differentiable density functional theory." Physical Review Letters 127.12 (2021): 126403.
+    
+    https://github.com/diffqc/dqc/blob/0fe821fc92cb3457fb14f6dff0c223641c514ddb/dqc/system/base_system.py
+    """
+
+    created_systems: Dict[str, DFTSystem] = {}
+
+    @classmethod
+    def create(cls, system: Dict) -> DFTSystem:
+        # create the system if it has not been created
+        # otherwise, return the previously created system
+
+        system_str = str(system)
+        if system_str not in cls.created_systems:
+            cls.created_systems[system_str] = DFTSystem(system)
+        return cls.created_systems[system_str]
+
+    def __init__(self, system: Dict):
+        super().__init__(system)
+
+        # caches
+        self._caches = {}
+
+    def get_dqc_system(self, pos_reqgrad: bool = False) -> BaseSystem:
+        """
+        This method converts the system dictionary to a DQC system and returns it. 
+        Parameters
+        ----------
+        pos_reqgrad: bool 
+            decides if the atomic position require gradient calculation. 
+        """
+        systype = self["type"]
+        if systype == "mol":
+            atomzs, atomposs = dqc.parse_moldesc(self["kwargs"]["moldesc"])
+            if pos_reqgrad:
+                atomposs.requires_grad_()
+            mol = Mol(**self["kwargs"])
+            return mol
+        else:
+            raise RuntimeError("Unknown system type: %s" % systype)
+
+    def get_cache(self, s: str):
+        return self._caches.get(s, None)
+
+    def set_cache(self, s: str, obj) -> None:
+        self._caches[s] = obj
+
+
+class DFTEntry(dict):
+    """
+    Handles creating and initialising DFTEntry objects from the dataset. This object contains information    about the various systems in the datapoint (atoms, molecules and ions) along with the ground truth
+    values.    
+    Note: Entry class should not be initialized directly, but created through
+    ``Entry.create``
+    """
+
+    created_entries: Dict[str, DFTEntry] = {}
+
+    @classmethod
+    def create(cls,
+               entry_dct: Union[Dict, DFTEntry],
+               device: torch.device,
+               dtype: torch.dtype = torch.double) -> DFTEntry:
+        if isinstance(entry_dct, DFTEntry):
+            return entry_dct
+
+        s = str(entry_dct)
+        if s not in cls.created_entries:
+            tpe = entry_dct["type"]
+            kwargs = {
+                "entry_dct": entry_dct,
+                "dtype": dtype,
+                "device": device,
+            }
+            obj = {
+                "ae": EntryAE,
+                "ie": EntryIE,
+                "dm": EntryDM,
+                "dens": EntryDens,
+                "force": EntryForce,
+            }[tpe](**kwargs)
```

Comment:
This is an arcane invocation! Took me 2 minutes to understand how it works. For maintainable code, let's just swap out to an explicit if/else series and avoid `kwargs` and instead use explicit constructor arguments